resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -115,6 +116,19 @@ private[spark] class ApplicationMaster(args: ApplicationMasterArguments) extends
     }
   }
 
+  if (isClusterMode) {
+    val nmHostPort = WebAppUtils.getWebAppBindURL(yarnConf, YarnConfiguration.NM_BIND_HOST,
```

Comment:
Can this be done above where the configuration is prepared? Then you wouldn't need to manually add this to the system properties since it would be done by the existing code.

Also it would be good to add a comment explaining why this is needed.